### PR TITLE
change selector return type

### PR DESCRIPTION
**Pull request by tienifr**
### Related Issues

GH_LINK

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by chiragsalian** [comment]
Is there an App issue to link to? Even if the change is good is there some issue explaining why this change is needed now?

**Comment by youssef-lr** [comment]
Context [here](https://github.com/Expensify/App/pull/36713/files#r1503618131) @chiragsalian, but I think it's good to run this by Callstack (or SWM) first.

**Comment by youssef-lr** [comment]
Asked [here](https://expensify.slack.com/archives/C0593T50UHG/p1709154506992979).

**Comment by chiragsalian** [comment]
Cool, thank you

File: lib/withOnyx.d.ts [hunk 1]
```diff
@@ -104,8 +104,8 @@ type CollectionMapping<TComponentProps, TOnyxProps, TOnyxProp extends keyof TOny
     CollectionBaseMapping<TOnyxKey> &
     (
         | BaseMappingKey<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey, OnyxCollection<KeyValueMapping[TOnyxKey]>>
-        | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey>
-        | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey>
+        | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, Partial<KeyValueMapping[TOnyxKey]>, TOnyxKey>
+        | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, Partial<KeyValueMapping[TOnyxKey]>, TOnyxKey>
```
Comment (fabioh8010) [inline lib/withOnyx.d.ts:108]:
```suggestion
        | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, Partial<KeyValueMapping[TOnyxKey]>, TOnyxKey>
        | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, Partial<KeyValueMapping[TOnyxKey]>, TOnyxKey>
```

@tienifr We have two issues here:
1. By changing `TOnyxProps[TOnyxProp]` to `Partial<KeyValueMapping[TOnyxKey]>` we lose all type equality check between the components' `OnyxProps` and `withOnyx` mapping, which means that we could type anything in the Onyx prop and the selector wouldn't be able to check if its return type matches with the corresponding Onyx prop.
2. Also, by doing `Partial<KeyValueMapping[TOnyxKey]>` you are limiting the selector's return to be an object with optional properties of that Onyx value, _which is not always the case_.
   
   When we have a combination of a collection key and a `selector`, the `withOnyx` will run `selector` function for each entry of that collection and build a collection object with the results of all `selector` executions, for example:

   ```tsx
   type ContactMethodDetailsPageOnyxProps = {
       policiesIDs: OnyxCollection<string>;
   };
   
   ...
   
   export default withOnyx<ContactMethodDetailsPageProps, ContactMethodDetailsPageOnyxProps>({
       policiesIDs: {
           key: ONYXKEYS.COLLECTION.POLICY,
           selector: (data) => data?.id ?? '', // this is valid and will make policies return a collection of ids
       },
   })(ContactMethodDetailsPage);
   ```

To improve these typings, my suggestion is:

1. Add this type in `lib/types.d.ts` which will make possible extracting the value of a Onyx collection:
   ```tsx
   ...
   type OnyxCollection<TOnyxValue> = OnyxEntry<Record<string, TOnyxValue | null>>;

   /**
    * Utility type to extract `TOnyxValue` from `OnyxCollection<TOnyxValue>`.
    */
   type ExtractOnyxCollectionValue<TOnyxCollection> = TOnyxCollection extends NonNullable<OnyxCollection<infer U>> ? U : never;
   ...
   ```
2. Use it in `CollectionMapping`:
   ```tsx
   /**
    * Represents the mapping options between an Onyx collection key without suffix and the component's prop with all its possibilities.
    */
   type CollectionMapping<TComponentProps, TOnyxProps, TOnyxProp extends keyof TOnyxProps, TOnyxKey extends CollectionKeyBase> = BaseMapping<TComponentProps, TOnyxProps> &
       CollectionBaseMapping<TOnyxKey> &
       (
           | BaseMappingKey<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey, OnyxCollection<KeyValueMapping[TOnyxKey]>>
           | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, ExtractOnyxCollectionValue<TOnyxProps[TOnyxProp]>, TOnyxKey>
           | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, ExtractOnyxCollectionValue<TOnyxProps[TOnyxProp]>, TOnyxKey>
       );
   ```

This way we can ensure the type equality checks will still work and get rid of these issues when using the `selector`. For your case [in this discussion](https://github.com/Expensify/App/pull/36713/files#discussion_r1502293136), final solution would be like this:

```tsx
type ContactMethodDetailsPageOnyxProps = {
    ...
    policies: OnyxCollection<Pick<Policy, 'id' | 'ownerAccountID'>>;
};

...

export default withOnyx<ContactMethodDetailsPageProps, ContactMethodDetailsPageOnyxProps>({
    ...
    policies: {
        key: ONYXKEYS.COLLECTION.POLICY,
        selector: (data) => ({id: data?.id ?? '', ownerAccountID: data?.ownerAccountID}),
    },
})(ContactMethodDetailsPage);
```

File: lib/withOnyx.d.ts [hunk 2]
```diff
@@ -93,8 +93,8 @@ type Mapping<TComponentProps, TOnyxProps, TOnyxProp extends keyof TOnyxProps, TO
     EntryBaseMapping<TOnyxKey> &
     (
         | BaseMappingKey<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey, OnyxEntry<KeyValueMapping[TOnyxKey]>>
-        | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProps[TOnyxProp], TOnyxKey>
-        | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProps[TOnyxProp], TOnyxKey>
+        | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, ExtractOnyxCollectionValue<TOnyxProps[TOnyxProp]>, TOnyxKey>
+        | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, ExtractOnyxCollectionValue<TOnyxProps[TOnyxProp]>, TOnyxKey>
```
Comment (fabioh8010) [inline lib/withOnyx.d.ts:97]:
```suggestion
        | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProps[TOnyxProp], TOnyxKey>
        | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProps[TOnyxProp], TOnyxKey>
```

You changed the wrong Mapping 😅

File: lib/withOnyx.d.ts [hunk 3]
```diff
@@ -104,8 +104,8 @@ type CollectionMapping<TComponentProps, TOnyxProps, TOnyxProp extends keyof TOny
     CollectionBaseMapping<TOnyxKey> &
     (
         | BaseMappingKey<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey, OnyxCollection<KeyValueMapping[TOnyxKey]>>
-        | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, Partial<KeyValueMapping[TOnyxKey]>, TOnyxKey>
-        | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, Partial<KeyValueMapping[TOnyxKey]>, TOnyxKey>
+        | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProps[TOnyxProp], TOnyxKey>
+        | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, TOnyxProps[TOnyxProp], TOnyxKey>
```
Comment (fabioh8010) [inline lib/withOnyx.d.ts:108]:
```suggestion
        | BaseMappingStringKeyAndSelector<TComponentProps, TOnyxProps, ExtractOnyxCollectionValue<TOnyxProps[TOnyxProp]>, TOnyxKey>
        | BaseMappingFunctionKeyAndSelector<TComponentProps, TOnyxProps, ExtractOnyxCollectionValue<TOnyxProps[TOnyxProp]>, TOnyxKey>
```

**Review by fabioh8010 — APPROVED** [review]

File: lib/types.d.ts [hunk 4]
```diff
@@ -183,6 +183,11 @@ type OnyxEntry<TOnyxValue> = TOnyxValue | null;
  */
 type OnyxCollection<TOnyxValue> = OnyxEntry<Record<string, TOnyxValue | null>>;
 
+/**
+ * * Utility type to extract `TOnyxValue` from `OnyxCollection<TOnyxValue>`.
+ */
```
Comment (blazejkustra) [inline lib/types.d.ts:188]: NAB:
```suggestion
/** Utility type to extract `TOnyxValue` from `OnyxCollection<TOnyxValue>` */
```

**Review by blazejkustra — APPROVED** [review]

**Comment by youssef-lr** [comment]
Thanks for working on this folks! May I ask if we can update the types to accept `props`? I'm seeing this error when passing the props to the selector, the props should contain the previously loaded keys

Screenshot 2024-03-01 at 13 23 41

**Comment by youssef-lr** [comment]
Hmm actually it doesn't work at all, not due to TS checks, but I see that we use `props` [here](https://github.com/Expensify/App/blob/28a1ccc40ad6345338de1951c02146304fb30cd5/src/pages/home/ReportScreen.js#L650), I wonder what's different?

**Comment by fabioh8010** [comment]
@youssef-lr Do you mind sharing the code that isn't working? I tested here and the types are correct.

Screenshot 2024-03-01 at 12 39 25

**Comment by youssef-lr** [comment]
Hm I think the changes in react-native-onyx are not being reflected for me locally, here's what I did:

1. I checked out this PR in `react-native-onyx`
2. I ran `npm run build:watch`
3. I checked out @tienifr's PR in App, deleted `.cache` folder in `node_modules`
4. I ran `npx link publish ~/Code/Expensidev/react-native-onyx`
5. I ran `npm run web`

Am I missing something?

**Comment by fabioh8010** [comment]
@youssef-lr After doing `npx link publish ~/Code/Expensidev/react-native-onyx`, try accessing `node_modules/react-native-onyx/dist/withOnyx.d.ts` and `node_modules/react-native-onyx/dist/index.d.ts` to ensure they are updated. Types might start working after this.

**Comment by youssef-lr** [comment]
They are updated..maybe it's just an issue with my editor. Thanks for checking!

**Comment by youssef-lr** [comment]
Any idea why the selector takes a single policy instead of all the policies from the collection? Probably out of scope here, but I was under the assumption the param of the selector should be all of the policies

**Comment by youssef-lr** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by fabioh8010** [comment]
@youssef-lr When we have this combination of a collection key and `selector`, the `selector` will actually be executed for each item of the collection, and then a collection is returned to the component. Take a look [at this function](https://github.com/Expensify/react-native-onyx/blob/main/lib/Onyx.js#L124) and where it's used in `lib/Onyx.js`.

**Comment by fabioh8010** [comment]
Currently, if you make the selector return `null` (I have to do that assertion due to a limitation in `withOnyx` typings), the result collection would be like this:

Screenshot 2024-03-01 at 13 22 57

```ts
{
    "policy_1576B20B2BA20523": {
        "isFromFullPolicy": false,
        "id": "1576B20B2BA20523",
        "name": "0.7574482937523088",
        "role": "admin",
        "type": "free",
        "owner": "0.3004884578151028",
        "ownerAccountID": 14357020,
        "outputCurrency": "EUR",
        "avatar": "",
        "employeeList": [],
        "isPolicyExpenseChatEnabled": true,
        "lastModified": 1701421196016165,
        "chatReportIDAnnounce": 1038301144060652,
        "chatReportIDAdmins": 5618866612197321
    },
    "policy_4EB3958A3E59A354": null,
    "policy_5D57E1CE8332729C": null,
    "policy_6C774C6E2CC5D5FD": null,
    "policy_9753CC55D5893F78": null,
    "policy_BBEDBB1D8A80E431": null,
    "policy_undefined": null,
    "policy_0": null,
    "policy_9FEECFE647F217AE": null
}
```

Which you could use it to filter the `null` values (in case you need the number of the records that were "selected").

Your suggestion makes sense but I don't know how it's the impact of it tbh 😅

**Comment by youssef-lr** [comment]
So here's a use case I'm thinking of, suppose we have a component that only needs a subset of a collection to subscribe to, we don't want re-renders if other items from that collection changes, I think only passing the items to the component would be beneficial here. Buuut, I suppose filtering out `null` policies and using `useMemo` would work the same. Thanks for chiming in!

**Pull request merged**